### PR TITLE
Fully add python typing.

### PR DESCRIPTION
**Pull request by coledie**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

New Behavior
----------------
* Examples folder is now type checked to help verify API is accurate.
* Add and updates types on all python/deepgnn folders: graph_engine, torch and tf.
* Refactor some functions.
* Query output type and forward input type is now QueryOutput instead of dict.

### CODE REVIEW

File: .github/actions/lint/action.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ runs:
     - run: brew install clang-format
       shell: bash
       if: runner.os == 'macOS'
-    - run: pip install wheel pre-commit==2.17.0 mypy==0.780 numpy==1.21.6
+    - run: pip install wheel pre-commit==2.17.0 mypy==0.971 numpy==1.21.6 torch==1.11.0 tensorflow==2.7.2
```
Comment (mortid0) [inline .github/actions/lint/action.yml:15]: why do we need to install tensorflow/torch for a linter?

File: .mypy.ini [hunk 2]
```diff
@@ -3,7 +3,7 @@ warn_return_any = False
 warn_unused_configs = True
 mypy_path = src/python/deepgnn
 
-files=src/python/deepgnn
+files=src/python/deepgnn,examples/*/*/*.py
```
Comment (mortid0) [inline .mypy.ini:6]: nit: use globbing to much every file in the repo `**/*.py`

File: examples/pytorch/__init__.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""DeepGNN model examples."""
```
Comment (mortid0) [inline examples/pytorch/__init__.py:3]: nit: """DeepGNN Pytorch model examples."""

File: examples/pytorch/gat/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""DeepGNN model examples."""
```
Comment (mortid0) [inline examples/pytorch/gat/__init__.py:3]: nit: """DeepGNN Pytorch GAT model."""

File: examples/pytorch/geometric/gat/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""DeepGNN model examples."""
```
Comment (mortid0) [inline examples/pytorch/geometric/gat/__init__.py:3]: nit: """DeepGNN GAT model implemented with Pytorch Geometric."""

Comment (mortid0) [reply examples/pytorch/geometric/gat/__init__.py:3]: same in other init files.

File: examples/pytorch/graphsage/main.py [hunk 6]
```diff
@@ -19,7 +19,7 @@
     SamplingStrategy,
     GraphEngineBackend,
 )
-from model import SupervisedGraphSage, UnSupervisedGraphSage
+from model import SupervisedGraphSage, UnSupervisedGraphSage  # type: ignore
```
Comment (mortid0) [inline examples/pytorch/graphsage/main.py:22]: why do we need to ignore types in imports?

File: .github/actions/lint/action.yml (hunk 1 above)
Comment (coledie) [reply .github/actions/lint/action.yml:15]: Otherwise it will ignore torch.Tensor and tf.Tensor types - it does not add any time to CI pre-commit it looks like.

File: examples/pytorch/graphsage/main.py (hunk 6 above)
Comment (coledie) [reply examples/pytorch/graphsage/main.py:22]: Mypy does not handle the repeated filenames and it is hard to get the imports right for each of the examples. The most important thing is how the examples use the trainer and GE to verify those types, not so much the model implementation is typed right.

File: .mypy.ini (hunk 2 above)
Comment (coledie) [reply .mypy.ini:6]: I've updated it to "src/python/deepgnn,examples" instead, globbing is hard since bazel need to be ignored.

File: .github/actions/lint/action.yml (hunk 1 above)
Comment (mortid0) [reply .github/actions/lint/action.yml:15]: pip install runs for 1 minute instead of 5 seconds with python 3.9 on ubuntu. Can you [cache it with a requirements.txt file](https://github.com/actions/cache/blob/main/examples.md#multiple-oss-in-a-workflow-with-a-matrix) and reference it in the doc?

File: .mypy.ini [hunk 7]
```diff
@@ -32,6 +32,9 @@ ignore_missing_imports = True
 [mypy-tensorflow.*]
 ignore_missing_imports = True
```
Comment (mortid0) [inline .mypy.ini:33]: Since we install tf/torch now, their missing imports should be removed, right?

**Review by mortid0 — APPROVED** [review]

File: .github/actions/lint/action.yml (hunk 1 above)
Comment (coledie) [reply .github/actions/lint/action.yml:15]: The downloads are cached: https://github.com/microsoft/DeepGNN/runs/7928528362?check_suite_focus=true#:~:text=Collecting%20torch%3D%3D1.11.0,67

File: .mypy.ini (hunk 2 above)
Comment (mortid0) [reply .mypy.ini]: use [exclude section](https://mypy.readthedocs.io/en/stable/config_file.html#confval-exclude)

Comment (coledie) [reply .mypy.ini]: I spent quite a while trying to get that to work and it seems to have weird behavior and I cannot get it to work. I think the order of the commands matter and maybe there are some overrides between commands or something that is not documented.

Comment (mortid0) [reply .mypy.ini]: Create an issue on the mypy github page and describe what you tried to do?

Comment (coledie) [reply .mypy.ini]: Right now files=src/python/deepgnn,examples which is not so bad. It is probably not worth the effort?

File: .mypy.ini (hunk 7 above)
Comment (coledie) [reply .mypy.ini:33]: It works on my local, but not on CI. I think this is a separate issue: https://github.com/microsoft/DeepGNN/runs/7963445411?check_suite_focus=true#:~:text=examples/pytorch/link_prediction/poolers.py%3A5%3A%20error%3A%20Cannot%20find%20implementation%20or%20library%20stub%20for%20module%20named%20%22torch%22

File: .mypy.ini (hunk 2 above)
Comment (mortid0) [reply .mypy.ini]: Well, you forgot to add `docs` folder, so it seems worth to use a correct pattern and exclude only bazel-bin folder.

File: .mypy.ini (hunk 7 above)
Comment (mortid0) [reply .mypy.ini:33]: Does your local python version match CI version?

Comment (coledie) [reply .mypy.ini:33]: Yes and same package versions

File: .mypy.ini (hunk 2 above)
Comment (coledie) [reply .mypy.ini]: docs does not have any code to type check. I dont think we should ever have code outside of src and examples to need a complex pattern.
